lib/multiplexer/proxyline.go

```diff
@@ -238,11 +286,61 @@ func ReadProxyLineV2(reader *bufio.Reader) (*ProxyLine, error) {
 		return nil, trace.BadParameter("unsupported protocol %x", header.Protocol)
 	}
 
+	// If there are more bytes left it means we've got TLVs
 	if header.Length > size {
-		if _, err := io.CopyN(io.Discard, reader, int64(header.Length-size)); err != nil {
+		tlvsBytes := &bytes.Buffer{}
+
+		if _, err := io.CopyN(tlvsBytes, reader, int64(header.Length-size)); err != nil {
 			return nil, trace.Wrap(err)
 		}
+
+		tlvs, err := UnmarshalTVLs(tlvsBytes.Bytes())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		ret.TLVs = tlvs
 	}
 
 	return &ret, nil
 }
+
+func UnmarshalTVLs(rawBytes []byte) ([]TLV, error) {
+	var tlvs []TLV
+	for i := 0; i < len(rawBytes); {
+		tlv := TLV{
+			Type: PP2Type(rawBytes[i]),
+		}
+		if len(rawBytes)-i < 2 {
+			return nil, ErrTruncatedTLV
+		}
+		tlvLen := int(binary.BigEndian.Uint16(rawBytes[i+1 : i+3]))
```

Comment:
nit: I think we can better express intent here by breaking this up a little bit more, to make it clearer what lengths we expect things to be.

I'd probably do this by introducing two new variables:

```go
lenStart := i+1
lenEnd := lenStart + 2
tlvLen := int(binary.BigEndian.Uint16(rawBytes[lenStart : lenEnd]))
```

I think this makes it much more explicit that we expect the length of the length (lol) to be 2 bytes.

Alternatively, we could continue to bump `i` after reading the type:

```go
i += 1
lenLength := 2
tlvLen := int(binary.BigEndian.Uint16(rawBytes[i : i+lenLength]))
i += lenLength
```

Reply:
A reader seems like a pretty sensible suggestion here, and would definitely make this section easier to read I would think.

Reply:
I tried with reader and I think it looked a bit clunky for this case. I've changed to Noah's first suggestion + added comments + added specific tests for `UnmarshalTLVs` + added fuzz tests.